### pip install sympy==1.13.3 scipy==1.15.2 numpy==2.2.4 matplotlib==3.10.1

# RECORDEMOS COMO DEFINIR UNA EDO DE PRIMER ORDEN

In [2]:
import numpy as np
import sympy as sp
import matplotlib.pyplot as plt

sp.init_printing(use_latex='mathjax') 

x = sp.Symbol('x', real=True)
y = sp.Function('y')


# METODO DE EULER

In [3]:
import matplotlib.pyplot as plt
import numpy as np

def Euler(f,x0,xn,y0,n): #nuestra función se llama Euler y sus datos de entrada son f,x0,xn,y0,n
    X = np.linspace(x0,xn,n+1)         
    Y = np.linspace(x0,xn,n+1)                 
    Y[0] = y0
    h=(xn-x0)/n
    for i in range(n):
        Y[i+1] = Y[i] + h*f(X[i],Y[i])
    return Y

# METODO RK4

In [52]:
def RK4(f, x0, xn, y0, n):
    X = np.linspace(x0,xn,n+1)
    Y = np.linspace(x0,xn,n+1)                 
    Y[0] = y0
    h = (xn-x0)/n     
    for i in range(n):
        K1=f(X[i],Y[i]) 
        K2=f(X[i]+h/2,Y[i]+(h/2)*K1) 
        K3=f(X[i]+h/2,Y[i]+(h/2)*K2) 
        K4=f(X[i]+h,Y[i]+h*K3)
        Y[i+1] = Y[i] +(h/6)*(K1+2*K2+2*K3+K4)
    return Y[-1]

# Y el calculo del error

valor absoluto entre la resta del valor original y el metodo aplicado, dividido por el valor original y luego multiplicada por 100

Si queremos que el resultado se muestre con más valores después de la coma, hacemos:

### Ejercicio 01

Considere el problema de valores iniciales:

$$ y'=\dfrac{y}{x}+x\mathrm{e}^{-\frac{y}{x}} \quad ; \quad y(1)=2.$$

* a) Utilice el método de Euler para encontrar una aproximación de $y(11)$ donde $h=1$.
* b) Usando Python resuelva la ecuación y determine si es una buena aproximación (haga un análisis del error obtenido).

# A)

In [5]:
def g(x,y): return (y/x) + x*sp.exp(-y/x)

In [6]:
euler1 = Euler(g,1,11,2,10)[-1]
euler1

np.float64(30.273285998556908)

# B)

In [7]:
f = (y(x)/x) + x*sp.exp(-y(x)/x)
sol = sp.dsolve(y(x).diff(x) - f, ics={y(1):2})
sol2 = sol.rhs
sol3 = sol2.subs(x,11).evalf()
sol3

31.4142515336177

In [8]:
np.abs((sol3 - euler1)/sol3)*100

3.63199974330060

### Ejercicio 03: 

La ecuación
$$\frac{dP}{dt}=kP\left(1-\frac{P}{N} \right)-a$$
representa unn modelo logístico de crecimiento poblacional con cosecha constante.

Para una cierta variedad de pez comestible, se considera $a=0.5$, $k=3$ y $N=10$, donde $N$ y $P(t)$ están medidas en miles.

Suponga que en el instante $t=0$ (medido en días) la población inicial de peces es $2600$.

* a) Use el método RK4 para estimar la población de peces al cabo de una semana.

* b) Estime la cantidad de peces diarios al cabo de un año completo. ¿La población de peces se estabiliza después de cierto tiempo? ¿Qué se espera de $P(t)$ si $t \to \infty$?

* c) La EDO resultante es autónoma. Utilice técnicas cualitativas para corroborar los resultados anteriores.

* d) Resuelva la EDO vía separación de variables, luego calcule la población al cabo de un año. Compruebe si sus resultados coinciden con los obtenidos anteriormente.

p(0)=2600 ics={y()}

In [26]:
a = 0.5
k = 3
N = 10

t = sp.Symbol('t', real=True)
P = sp.Function('P')


A)

In [53]:
# def RK4(f, x0, xn, y0, n)

def h(x,y): return (k*y*(1-(y/N)) - a)

# en una semana hay 7 dias

# debido a un problema con overflow, aumentamos la cantidad de iteraciones de 10 a 10000
RK4(h, 0, 7, 2600, 10000)

np.float64(9.830458930271815)

# B)

In [54]:
# P(t) bajarlo a 2.6
RK4(h, 0, 365, 2.6, 10000)

np.float64(9.830458915396472)

In [75]:
RK4(h, 0, 5*365, 2.6, 10000)

np.float64(9.830458915396479)

In [76]:
RK4(h, 0, 10*365, 2.6, 10000)

np.float64(9.83045891539648)

c)

In [78]:
sp.solve(k*x*(1-(x/N)) - a)[-1]

9.83045891539648

# D)
## me dio en la integral por parte del y, esto: 💀💀💀💀💀💀💀💀💀💀💀💀
### Muy probable es tambien el porque tuimos overflow

$$
-\frac{5 \ln\left(\frac{\left|3y + \frac{-2 \sqrt{210} - 30}{2}\right|}{\left|3y + \frac{2 \sqrt{210} - 30}{2}\right|}\right)}{\sqrt{210}}
$$


# Mas Ejercicios

## Ejercicio 01
$$\frac{dy}{dx} = \frac{y}{x} + xe^{-\frac{y}{x}}, \quad y(1) = 2$$

* A) Método de Euler con h = 1 para y(11)
* B) Solución exacta y porcentaje de error

---

## Ejercicio 02
$$\frac{dy}{dx} = 0.2xy, \quad y(0) = 1$$

* A) RK4 con h = 1 para y(4)
* B) Solución exacta y porcentaje de error

---

## Ejercicio 03
$$\frac{dP}{dt} = 3P\left(1-\frac{P}{10}\right) - 0.5, \quad P(0)=2.6$$

* A) RK4 para P(7 días)
* B) Estabilidad a largo plazo (t->oo)
* C) Análisis cualitativo

---

## Ejercicio 04
$$\frac{dy}{dx} = \frac{y}{x} + x\cos\left(\frac{y}{x}\right), \quad y(1)=0$$

* A) Euler con h = 0.5 para y(3)
* B) Porcentaje de error

---

## Ejercicio 05
$$\frac{dy}{dx} = 1 + \frac{y}{x}, \quad y(1)=2$$

* A) RK4 con h = 0.25 para y(2)
* B) Comparación exacta

---

## Ejercicio 06
$$\frac{dv}{dt} = 9.8 - 0.1v, \quad v(0)=0$$

* A) RK4 con h = 0.1 para v(2)
* B) Velocidad límite

---

## Ejercicio 07
$$\frac{dP}{dt} = 2P\left(1-\frac{P}{100}\right) - \frac{P^2}{200}, \quad P(0)=10$$

* A) Euler con h = 1 para P(5)
* B) Puntos de equilibrio

---

## Ejercicio 08
$$\frac{dy}{dx} = y - x^2 + 1, \quad y(0)=0.5$$

* A) RK4 con h = 0.2 para y(1)
* B) Euler con h = 0.2 para y(1)


---

## Ejercicio 09
$$\frac{dy}{dx} = \frac{e^{-y}}{x}, \quad y(1)=0$$

* A) Solución exacta
* B) Euler con h = 0.5 para y(3)

---

## Ejercicio 10
$$\frac{dT}{dt} = -0.1(T-20), \quad T(0)=100$$

* A) RK4 con h = 1 para T(10)
* B) Solución exacta